File: !FZOJ/C1543/fib_bot.py
import sys

input = sys.stdin.read
data = input().split()

index = 0
id_ = int(data[index])
index += 1
n = int(data[index])
index += 1
m = int(data[index])
index += 1

MOD = 1 << 64  # 2**64, unsigned long long overflow is mod 2^64

class Matrix:
    def __init__(self, a=0, b=0, c=0, d=0):
        self.a = a
        self.b = b
        self.c = c
        self.d = d

    def __mul__(self, other):
        return Matrix(
            (self.a * other.a + self.b * other.c) & (MOD - 1),
            (self.a * other.b + self.b * other.d) & (MOD - 1),
            (self.c * other.a + self.d * other.c) & (MOD - 1),
            (self.c * other.b + self.d * other.d) & (MOD - 1)
        )

# Precompute fib[1..n+1]
fib = [0] * (n + 2)
fib[1] = 1
fib[2] = 1
for i in range(3, n + 2):
    fib[i] = (fib[i - 1] + fib[i - 2]) & (MOD - 1)

class Node:
    def __init__(self):
        self.sum = 0          # 区间和
        self.sqsum = 0        # 区间平方和
        self.fa = 0            # 斐波那契线性表示系数 F_{pos-1}
        self.fb = 0            # 斐波那契线性表示系数 F_{pos}
        self.lazy_a = 0
        self.lazy_b = 0

tree = [Node() for _ in range(4 * (n + 1))]

def push_down(p, len_):
    if tree[p].lazy_a == 0 and tree[p].lazy_b == 0:
        return
    left_len = len_ // 2
    right_len = len_ - left_len
    l = p * 2
    r = p * 2 + 1

    # 左子区间
    la = tree[p].lazy_a
    lb = tree[p].lazy_b
    old_sum_l = tree[l].sum
    old_sq_l = tree[l].sqsum
    add_sum_l = (la * fib[left_len + 1] + lb * fib[left_len]) & (MOD - 1)
    add_sq_l = (
        la * la * fib[left_len * 2 + 1] +
        lb * lb * fib[left_len * 2 - 1] +
        2 * la * lb * fib[left_len * 2]
    ) & (MOD - 1)
    tree[l].sum = (old_sum_l + add_sum_l) & (MOD - 1)
    tree[l].sqsum = (old_sq_l + add_sq_l + 2 * ((old_sum_l * add_sum_l) & (MOD - 1))) & (MOD - 1)

    # 合并到左子的lazy
    new_fa_l = (tree[l].fa + lb) & (MOD - 1)
    new_fb_l = (tree[l].fb + la + lb) & (MOD - 1)
    tree[l].lazy_a = (tree[l].lazy_a + la) & (MOD - 1)
    tree[l].lazy_b = (tree[l].lazy_b + lb) & (MOD - 1)
    tree[l].fa = new_fa_l
    tree[l].fb = new_fb_l

    if right_len > 0:
        # 右子区间
        ra = (la * fib[2] + lb * fib[3]) & (MOD - 1)  # F_{right_start-1}
        rb = (la * fib[3] + lb * fib[4]) & (MOD - 1)  # F_{right_start}
        old_sum_r = tree[r].sum
        old_sq_r = tree[r].sqsum
        add_sum_r = (ra * fib[right_len + 1] + rb * fib[right_len]) & (MOD - 1)
        add_sq_r = (
            ra * ra * fib[right_len * 2 + 1] +
            rb * rb * fib[right_len * 2 - 1] +
            2 * ra * rb * fib[right_len * 2]
        ) & (MOD - 1)
        tree[r].sum = (old_sum_r + add_sum_r) & (MOD - 1)
        tree[r].sqsum = (old_sq_r + add_sq_r + 2 * ((old_sum_r * add_sum_r) & (MOD - 1))) & (MOD - 1)

        new_fa_r = (tree[r].fa + rb) & (MOD - 1)
        new_fb_r = (tree[r].fb + ra + rb) & (MOD - 1)
        tree[r].lazy_a = (tree[r].lazy_a + ra) & (MOD - 1)
        tree[r].lazy_b = (tree[r].lazy_b + rb) & (MOD - 1)
        tree[r].fa = new_fa_r
        tree[r].fb = new_fb_r

    tree[p].lazy_a = 0
    tree[p].lazy_b = 0

def push_up(p):
    l = p * 2
    r = p * 2 + 1
    tree[p].sum = (tree[l].sum + tree[r].sum) & (MOD - 1)
    tree[p].sqsum = (tree[l].sqsum + tree[r].sqsum) & (MOD - 1)

def build(p, cl, cr):
    if cl == cr:
        tree[p].sum = 0
        tree[p].sqsum = 0
        tree[p].fa = fib[cl - 1] if cl >= 2 else (1 if cl == 1 else 0)
        tree[p].fb = fib[cl]
        return
    mid = (cl + cr) // 2
    build(p * 2, cl, mid)
    build(p * 2 + 1, mid + 1, cr)
    push_up(p)
    tree[p].fa = fib[cl - 1] if cl >= 2 else (1 if cl == 1 else 0)
    tree[p].fb = fib[cl]

def update(p, cl, cr, l, r):
    if l <= cl and cr <= r:
        len_ = cr - cl + 1
        la = fib[cl - l + 1]
        lb = fib[cl - l + 2]
        old_sum = tree[p].sum
        old_sq = tree[p].sqsum
        add_sum = (la * fib[len_ + 1] + lb * fib[len_]) & (MOD - 1)
        add_sq = (
            la * la * fib[len_ * 2 + 1] +
            lb * lb * fib[len_ * 2 - 1] +
            2 * la * lb * fib[len_ * 2]
        ) & (MOD - 1)
        tree[p].sum = (old_sum + add_sum) & (MOD - 1)
        tree[p].sqsum = (old_sq + add_sq + 2 * ((old_sum * add_sum) & (MOD - 1))) & (MOD - 1)
        tree[p].lazy_a = (tree[p].lazy_a + la) & (MOD - 1)
        tree[p].lazy_b = (tree[p].lazy_b + lb) & (MOD - 1)
        new_fa = (tree[p].fa + lb) & (MOD - 1)
        new_fb = (tree[p].fb + la + lb) & (MOD - 1)
        tree[p].fa = new_fa
        tree[p].fb = new_fb
        return
    push_down(p, cr - cl + 1)
    mid = (cl + cr) // 2
    if l <= mid:
        update(p * 2, cl, mid, l, r)
    if r > mid:
        update(p * 2 + 1, mid + 1, cr, l, r)
    push_up(p)

def query(p, cl, cr, l, r):
    if l <= cl and cr <= r:
        return tree[p].sqsum
    push_down(p, cr - cl + 1)
    mid = (cl + cr) // 2
    ans = 0
    if l <= mid:
        ans = (ans + query(p * 2, cl, mid, l, r)) & (MOD - 1)
    if r > mid:
        ans = (ans + query(p * 2 + 1, mid + 1, cr, l, r)) & (MOD - 1)
    return ans

build(1, 1, n)

ans_list = []
for _ in range(m):
    opt = int(data[index])
    index += 1
    l = int(data[index])
    index += 1
    r = int(data[index])
    index += 1
    if opt == 1:
        update(1, 1, n, l, r)
    else:
        res = query(1, 1, n, l, r)
        ans_list.append(str(res))

print('\n'.join(ans_list))

File: !FZOJ/C1543/fib_Bot.cpp
#include <bits/stdc++.h>
using namespace std;

using ull = unsigned long long;

int main() {
    int id, n, m;
    cin >> id >> n >> m;
    vector<ull> fib(n + 2, 0);
    fib[0] = 0;
    fib[1] = 1;
    fib[2] = 1;
    for (int i = 3; i <= n + 1; ++i) {
        fib[i] = fib[i - 1] + fib[i - 2];
    }

    struct Node {
        ull sum = 0;
        ull sqsum = 0;
        ull fa = 0;
        ull fb = 0;
        ull lazy_a = 0;
        ull lazy_b = 0;
    };

    vector<Node> tree(4 * (n + 1));

    auto get_s = [&fib](int m) -> ull {
        if (m <= 0) return 0;
        ull f = fib[m + 1];
        ull sq = f * f;
        if (m % 2 == 0) sq -= 1;
        return sq;
    };

    auto push_up = [&tree](int p) {
        tree[p].sum = tree[p * 2].sum + tree[p * 2 + 1].sum;
        tree[p].sqsum = tree[p * 2].sqsum + tree[p * 2 + 1].sqsum;
    };

    auto build = [&] (auto &self, int p, int cl, int cr) -> void {
        tree[p].fa = fib[cl - 1];
        tree[p].fb = fib[cl];
        if (cl == cr) return;
        int mid = (cl + cr) / 2;
        self(self, p * 2, cl, mid);
        self(self, p * 2 + 1, mid + 1, cr);
    };

    build(build, 1, 1, n);

    auto push_down = [&] (int p, int cl, int cr) -> void {
        ull la = tree[p].lazy_a;
        ull lb = tree[p].lazy_b;
        if (la == 0 && lb == 0) return;
        int mid = (cl + cr) / 2;
        int left_len = mid - cl + 1;
        int right_len = cr - mid;
        int l = p * 2, r = p * 2 + 1;

        // left
        ull add_sum_l = fib[left_len + 1] * (la + lb) + fib[left_len] * lb - (la + lb);
        ull old_sum_l = tree[l].sum;
        tree[l].sum = old_sum_l + add_sum_l;
        ull add_sq_l = la * la * fib[left_len - 1] * fib[left_len] + lb * lb * fib[left_len] * fib[left_len + 1] + 2 * la * lb * get_s(left_len - 1);
        tree[l].sqsum = tree[l].sqsum + add_sq_l + 2 * old_sum_l * add_sum_l;
        tree[l].lazy_a += la;
        tree[l].lazy_b += lb;
        tree[l].fa += lb;
        tree[l].fb += la + lb;

        if (right_len > 0) {
            // right
            ull ra = la * fib[left_len - 1] + lb * fib[left_len];
            ull rb = la * fib[left_len] + lb * fib[left_len + 1];
            ull add_sum_r = fib[right_len + 1] * (ra + rb) + fib[right_len] * rb - (ra + rb);
            ull old_sum_r = tree[r].sum;
            tree[r].sum = old_sum_r + add_sum_r;
            ull add_sq_r = ra * ra * fib[right_len - 1] * fib[right_len] + rb * rb * fib[right_len] * fib[right_len + 1] + 2 * ra * rb * get_s(right_len - 1);
            tree[r].sqsum = tree[r].sqsum + add_sq_r + 2 * old_sum_r * add_sum_r;
            tree[r].lazy_a += ra;
            tree[r].lazy_b += rb;
            tree[r].fa += rb;
            tree[r].fb += ra + rb;
        }

        tree[p].lazy_a = 0;
        tree[p].lazy_b = 0;
    };

    auto update = [&] (auto &self, int p, int cl, int cr, int ql, int qr) -> void {
        if (ql <= cl && cr <= qr) {
            ull la = fib[cl - ql];
            ull lb = fib[cl - ql + 1];
            ull len_ = cr - cl + 1;
            ull old_sum = tree[p].sum;
            ull add_sum = fib[len_ + 1] * (la + lb) + fib[len_] * lb - (la + lb);
            tree[p].sum = old_sum + add_sum;
            ull add_sq = la * la * fib[len_ - 1] * fib[len_] + lb * lb * fib[len_] * fib[len_ + 1] + 2 * la * lb * get_s(len_ - 1);
            tree[p].sqsum = tree[p].sqsum + add_sq + 2 * old_sum * add_sum;
            tree[p].lazy_a += la;
            tree[p].lazy_b += lb;
            tree[p].fa += lb;
            tree[p].fb += la + lb;
            return;
        }
        push_down(p, cl, cr);
        int mid = (cl + cr) / 2;
        if (ql <= mid) self(self, p * 2, cl, mid, ql, qr);
        if (qr > mid) self(self, p * 2 + 1, mid + 1, cr, ql, qr);
        push_up(p);
    };

    auto query = [&] (auto &self, int p, int cl, int cr, int ql, int qr) -> ull {
        if (ql <= cl && cr <= qr) return tree[p].sqsum;
        push_down(p, cl, cr);
        int mid = (cl + cr) / 2;
        ull ans = 0;
        if (ql <= mid) ans += self(self, p * 2, cl, mid, ql, qr);
        if (qr > mid) ans += self(self, p * 2 + 1, mid + 1, cr, ql, qr);
        return ans;
    };

    vector<ull> ans_list;
    for (int i = 0; i < m; ++i) {
        int opt, l, r;
        cin >> opt >> l >> r;
        if (opt == 1) {
            update(update, 1, 1, n, l, r);
        } else {
            ull res = query(query, 1, 1, n, l, r);
            ans_list.push_back(res);
        }
    }

    for (auto &res : ans_list) {
        cout << res << endl;
    }
    return 0;
}